### PR TITLE
Fcrepo 3919 - Make solr endpoint credentials configurable

### PR DESCRIPTION
**Pull request by Surfrdan**
**Make solr endpoint credentials configurable**
* * *

**JIRA Ticket**: https://fedora-repository.atlassian.net/jira/software/c/projects/FCREPO/issues/FCREPO-3919?filter=myopenissues

# What does this Pull Request do?
Adds configurable username and password properties for HTTP Basic Auth when indexing Solr documents into an auth protected Solr instance

# How should this be tested?

* Configure a Solr instance with HTTP Basic Auth enabled https://solr.apache.org/guide/8_1/basic-authentication-plugin.html
* Set properties solr.username and solr.password according to the documentation 
* Test solr indexing endpoint to confirm that authentication works for a set of valid credentials and fails for invalid credentials

# Interested parties
Tag (@ mention) interested parties or, if unsure, @fcrepo/committers

### CODE REVIEW

File: fcrepo-indexing-solr/src/main/java/org/fcrepo/camel/indexing/solr/SolrRouter.java [hunk 1]
```diff
@@ -180,6 +187,12 @@ public void configure() throws Exception {
                 .removeHeaders("CamelHttp*")
                 .setHeader(HTTP_METHOD).constant("POST")
                 .setHeader(HTTP_QUERY).simple("commitWithin=" + config.getCommitWithin())
+                .choice()
+                .when(useSolrAuth)
+                    .process(new AddBasicAuthProcessor(solrUsername, solrPassword))
+                    .log(LoggingLevel.DEBUG, logger, "Authenticating solr with: " + solrUsername + ":" + solrPassword)
+                .otherwise()
+                    .log(LoggingLevel.DEBUG, logger, "No Solr Auth provided")
```
Comment (whikloj) [inline fcrepo-indexing-solr/src/main/java/org/fcrepo/camel/indexing/solr/SolrRouter.java:195]: This `choice()` block is unnecessary as the `AddBasicAuthProcessor()` already checks that the provided username is not blank before applying the header to the request. See in the [TriplestoreRouter](https://github.com/fcrepo-exts/fcrepo-camel-toolbox/blob/main/fcrepo-indexing-triplestore/src/main/java/org/fcrepo/camel/indexing/triplestore/TriplestoreRouter.java#L133-L134) and the actual [AddBasicAuthProcessor](https://github.com/fcrepo-exts/fcrepo-camel-toolbox/blob/main/fcrepo-camel-common/src/main/java/org/fcrepo/camel/common/processor/AddBasicAuthProcessor.java#L36)

**Review by whikloj — CHANGES_REQUESTED** [review]
Might also be good to have a test to check that the Header is being passed along when the appropriate variables are set.

File: fcrepo-indexing-solr/src/main/java/org/fcrepo/camel/indexing/solr/SolrRouter.java (hunk 1 above)
Comment (whikloj) [reply fcrepo-indexing-solr/src/main/java/org/fcrepo/camel/indexing/solr/SolrRouter.java:195]: Due to the above not working, I'd say to just remove this `choice` and leave it as 
```java
.process(new AddBasicAuthProcessor(solrUsername, solrPassword))
.log(LoggingLevel.DEBUG, logger, "Authenticating solr with: " + solrUsername)
```
Also removing the password so we don't log it.

**Review by whikloj — CHANGES_REQUESTED** [review]
It took a bit to get a `security.json` block working but once it did then it appears that with `"blockUnknown": true` I could not send records.

To duplicate my infrastructure I used this repository, I checked out the PR
```bash
> git fetch origin refs/pull/200/merge
From https://github.com/fcrepo-exts/fcrepo-camel-toolbox
 * branch              refs/pull/200/merge -> FETCH_HEAD

> git checkout FETCH_HEAD
HEAD is now at 1681ce19 Merge 77d1bc81f081e3cd8718a17f173c6df6900acdc9 into c6f5467ba430b44f960e1c5ea59bf05d83d44df3
```
then I built it with `mvn clean install`

Then in the docker-compose directory I add a file `security.json` that contained
```json
{
"authentication":{
   "blockUnknown": true,
   "class":"solr.BasicAuthPlugin",
   "credentials":{
       "solr": "IV0EHq1OnNrj6gvRCwvFwTrZ1+z1oBbnQdiVC3otuq0= Ndd7LKvVBAaZIF0QAVi1ekCfAJXr1GGfLtRUXhgrF8c=",
       "test-user": "Q41CCgahKHBbz2uxCgtBlWhyCSUAwxP4o7/TqzYWrTg= t6TcXYqbQEWdQhDv2XKXDIDUyHu/KTIJru4XzZUYpgk="
   },
   "realm":"My Solr users",
   "forwardCredentials": false
},
"authorization":{
   "class":"solr.RuleBasedAuthorizationPlugin",
   "permissions": [
      {
         "name":"all",
         "role":"admin"
      }
   ],
   "user-role": {
      "solr": "admin",
      "test-user": "admin"
   }
}}
```
Then I modified the `docker-compose.yml` to have
```
diff --git a/docker-compose/docker-compose.yml b/docker-compose/docker-compose.yml
index 8b7850d0..f153ee70 100644
--- a/docker-compose/docker-compose.yml
+++ b/docker-compose/docker-compose.yml
@@ -15,6 +15,8 @@ services:
     command:
       - solr-precreate
       - fcrepo
+    volumes:
+      - "./security.json:/var/solr/data/security.json"
 
   fuseki:
     image: atomgraph/fuseki
@@ -25,7 +27,10 @@ services:
       - /fcrepo
       
   camel-toolbox:
-    image: fcrepo/fcrepo-camel-toolbox:latest
+    build:
+      context: ../
+      dockerfile: Dockerfile
     volumes:
       - "./camel-toolbox-config:/config"
     ports:
```
Lastly I ran `docker-compose up --build -d` and once it had built and raised the containers I used the `test-user:solrRocks` credentials to access the HTML admin site to confirm the credentials work.

File: fcrepo-indexing-solr/src/main/java/org/fcrepo/camel/indexing/solr/SolrRouter.java [hunk 2]
```diff
@@ -58,7 +61,10 @@ public void configure() throws Exception {
         ns.add("indexing", "http://fedora.info/definitions/v4/indexing#");
         ns.add("ldp", "http://www.w3.org/ns/ldp#");
 
-
+        final String solrUsername = config.getSolrUsername();
+        final String solrPassword = config.getSolrPassword();
+        final Predicate useSolrAuth = PredicateBuilder.constant(
+                "true".equals(!StringUtils.isEmpty(solrUsername) && !StringUtils.isEmpty(solrPassword)));
```
[the remark this answers is not in the document]
Comment (Surfrdan) [reply fcrepo-indexing-solr/src/main/java/org/fcrepo/camel/indexing/solr/SolrRouter.java:67]: Yeh I'm afraid the mists of time have caused me to forget what I was aiming for here. I seem to recall trying to get around not being able to logically test non-camel variables, but its unnecessary code anyway. The headers will simply be ignored if sold isn't configured for auth. I'll remove this

File: docker-compose/camel-toolbox-config/configuration.properties [hunk 3]
```diff
@@ -1,10 +1,11 @@
 fcrepo.baseUrl=http://fcrepo:8080/fcrepo/rest
-fcrepo.authHost=fcrepo
+fcrepo.authHost=localhost
```
Comment (whikloj) [inline docker-compose/camel-toolbox-config/configuration.properties:2]: This needs to be `fcrepo` or authentication against Fedora does not work.

**Comment by whikloj** [comment]
This PR works, has a bunch of already committed changes (I'm guessing the Github branch update button) but that also raised an issue with the configuration file noted above. If you could fix that then we can merge this it should be ok (🤞 )

**Pull request merged**